apiserver/facades/client/application/application.go

```diff
@@ -908,6 +896,11 @@ func (api *APIBase) SetCharm(args params.ApplicationSetCharm) error {
 	if err := api.checkCanWrite(); err != nil {
 		return err
 	}
+
+	if args.CharmOrigin != nil && !corecharm.CharmHub.Matches(args.CharmOrigin.Source) && !corecharm.Local.Matches(args.CharmOrigin.Source) {
```

Comment:
It might be useful to have a single point of check (like a util function or something) for validity (of `CharmOrigin`) -- this check is a duplicate of L277--, I imagine things might get messy if we ever add another source in the future. Besides that check might be useful somewhere else as well.

Reply:
Nice catch @cderici, I realized with the new function there was a bug in the logic, if CharmOrigin is nil, you can't print the source.